### PR TITLE
8290401: Support dump all phases and print nodes in ascending order of index

### PR DESCRIPTION
**Pull request by caojoshua**
* Sort nodes by index (`S` flag)
* increase max node distance to maximum number of nodes
* `DEBUG` phase for `-XX:CompileCommand=PrintIdealPhase` is currently unused. We propose to use `DEBUG` phase for printing all phases. Alternatively, we can add a new `ALL` phase. Seeking opinions from the community.

Thanks Xin Liu for working with me on this.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290401](https://bugs.openjdk.org/browse/JDK-8290401): Support dump all phases and print nodes in ascending order of index

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10371/head:pull/10371` \
`$ git checkout pull/10371`

Update a local copy of the PR: \
`$ git checkout pull/10371` \
`$ git pull https://git.openjdk.org/jdk pull/10371/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10371`

View PR using the GUI difftool: \
`$ git pr show -t 10371`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10371.diff">https://git.openjdk.org/jdk/pull/10371.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/phasetype.hpp [hunk 1]
```diff
@@ -168,6 +168,8 @@ class PhaseNameValidator {
         // strncpy always writes len characters. If the source string is shorter, the function fills the remaining bytes with NULLs.
         strncpy(_bad, *iter, len);
         _valid = false;
+      } else if (PHASE_DEBUG == cpt) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:171]: I don't think abusing the unused `PHASE_DEBUG` in this way is a good and clean approach. I'd rather add a new enum entry `PHASE_ALL` or something like that in `CompilerPhaseType` as you have proposed.

**Comment by eme64** [comment]
Updated `dump` with sorting makes sense, as you say in [JDK-8290401](https://bugs.openjdk.org/browse/JDK-8290401) this allows easier to take a diff.
Adding in the `tty->print_cr("AFTER: %s", phase_name);` is nice for human reader, and if there is an automatic tool that reads this it can hopefully easily be adapted.
I would recommend creating a new flag `PHASE_ALL`, this is more understandable.
`PHASE_DEBUG` is used in `igv_print_default` (for debugger, call this any time to take snapshot and send it to IGV for visualization) - I think it would be confusing to reuse this for something else.

**Review by chhagedorn — CHANGES_REQUESTED** [review]

**Comment by caojoshua** [comment]
New commits changes use of `PHASE_DEBUG` -> `PHASE_ALL`

File: src/hotspot/share/opto/phasetype.hpp [hunk 2]
```diff
@@ -168,6 +169,8 @@ class PhaseNameValidator {
         // strncpy always writes len characters. If the source string is shorter, the function fills the remaining bytes with NULLs.
         strncpy(_bad, *iter, len);
         _valid = false;
+      } else if (PHASE_ALL == cpt) {
+        mask = ~((uint64_t)0);
```
Comment (navyxliu) [inline src/hotspot/share/opto/phasetype.hpp:173]: Can we also change this from uint64_t to UINT64_C for the consistent style?

Comment (caojoshua) [reply src/hotspot/share/opto/phasetype.hpp]: done

**Review by navyxliu — APPROVED** [review]
LGTM. I am not a reviewer. Still need other reviewers to approve it.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Pull request closed without merging**